### PR TITLE
SONARJAVA-1811 Lookup hierarchy for sam resolution

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -893,21 +893,55 @@ private static JavaType getPrimitive(JavaType primitiveOrWrapper) {
...
+      return Optional.empty();
+    }
+    Optional<List<JavaType>> javaTypes = type.symbol().memberSymbols().stream()
+      .filter(Resolve::isAbstractMethod).findFirst()
+      .map(s -> ((MethodJavaType) ((JavaSymbol.MethodJavaSymbol) s).type).argTypes);
+
+    if(!javaTypes.isPresent()) {
+      javaTypes = findSamMethodArgsRecursively(type.symbol().superClass());
+      if(!javaTypes.isPresent()) {
+        javaTypes = type.symbol().interfaces().stream().map(this::findSamMethodArgsRecursively).filter(Optional::isPresent).map(Optional::get).findFirst();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 2]
```diff
@@ -893,21 +893,55 @@ private static JavaType getPrimitive(JavaType primitiveOrWrapper) {
...
+      return Optional.empty();
+    }
+    Optional<List<JavaType>> result = type.symbol().memberSymbols().stream()
+      .filter(Resolve::isAbstractMethod).findFirst()
+      .map(s -> ((MethodJavaType) ((JavaSymbol.MethodJavaSymbol) s).type).argTypes);
+
+    if(!result.isPresent()) {
+      result = findSamMethodArgsRecursively(type.symbol().superClass());
+      if(!result.isPresent()) {
+        result = type.symbol().interfaces().stream().map(this::findSamMethodArgsRecursively).filter(Optional::isPresent).map(Optional::get).findFirst();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

Comment (tvsonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: @benzonico What's the "Unknown method" warning here?

Comment (tvsonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: This line should be formatted as the other streamed calls

File: java-frontend/src/test/files/sym/references/MethodCall.java [hunk 3]
```diff
@@ -342,3 +342,21 @@ void plop() {
...
 }
+class OUTER {
+  class A {
+
+    void bar() {
+      B<Integer> b = new B<>();
+      b.foo(this::add);
+    }
+
+    private Integer add(Integer a, Integer b) { // The method is not resolved
```
Comment (tvsonar) [inline java-frontend/src/test/files/sym/references/MethodCall.java]: Remove comment

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java (hunk 2 above)
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: @tvsonar this warning comes from Dory and is raised when we fail to resolve a method (revealing a flaw in semantic)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 4]
```diff
@@ -893,21 +893,59 @@ private static JavaType getPrimitive(JavaType primitiveOrWrapper) {
...
+      .map(s -> ((MethodJavaType) ((JavaSymbol.MethodJavaSymbol) s).type).argTypes);
+
+    if(!result.isPresent()) {
+      result = findSamMethodArgsRecursively(type.symbol().superClass());
+      if(!result.isPresent()) {
+        result = type.symbol().interfaces().stream()
+          .map(this::findSamMethodArgsRecursively)
+          .filter(Optional::isPresent)
+          .map(Optional::get)
+          .findFirst();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Unknown method [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java (hunk 2 above)
Comment (tvsonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: Which method is the unresolved one here? Isn't this resolution issue resolved already? Shouldn't we just update the java plugin on Dory?

**Pull request merged**